File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Slovakia.m
//
//  UIImage+Slovakia.m
//  CountriesFlags
//

#import "Countries/UIImage+Slovakia.h"

@implementation UIImage (Slovakia)

+ (UIImage *)flagSlovakiaWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:37.0f/255.0f
                                         green:74.0f/255.0f
                                          blue:165.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:237.0f/255.0f
                                        green:28.0f/255.0f
                                         blue:36.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorWhite setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 3.0f, size.width, size.height / 3.0f));
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 3.0f * 2.0, size.width, size.height / 3.0f));
    
    [colorWhite setStroke];
    [colorRed setFill];
    UIBezierPath* path = [UIBezierPath bezierPath];
    // draw shield
    [path moveToPoint:CGPointMake(size.width / 90.0f * 15.0f, size.height / 60.0f * 14.5f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 14.5f, size.height / 60.0f * 28.5f)
            controlPoint1:CGPointMake(size.width / 90.0f * 14.7f, size.height / 60.0f * 18.0f)
            controlPoint2:CGPointMake(size.width / 90.0f * 14.7f, size.height / 60.0f * 20.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 27.0f, size.height / 60.0f * 45.5f)
            controlPoint1:CGPointMake(size.width / 90.0f * 14.8f, size.height / 60.0f * 36.3f)
            controlPoint2:CGPointMake(size.width / 90.0f * 18.1f, size.height / 60.0f * 40.8f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 39.5f, size.height / 60.0f * 28.5f)
            controlPoint1:CGPointMake(size.width / 90.0f * 40.1f, size.height / 60.0f * 38.8f)
            controlPoint2:CGPointMake(size.width / 90.0f * 39.0f, size.height / 60.0f * 31.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 38.96f, size.height / 60.0f * 14.5f)
            controlPoint1:CGPointMake(size.width / 90.0f * 39.4f, size.height / 60.0f * 20.0f)
            controlPoint2:CGPointMake(size.width / 90.0f * 39.3f, size.height / 60.0f * 18.0f)];
    [path addLineToPoint:CGPointMake(size.width / 90.0f * 15.0f, size.height / 60.0f * 14.5f)];
    [path closePath];
    path.lineWidth = size.width / 90.0f * (9.0f / 10.0f);
    [path fill];
    [path stroke];
    
    // draw central line
    [colorWhite setFill];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 90.0f * 26.0f, size.height / 60.0f * 40.0f)];
    [path addLineToPoint:CGPointMake(size.width / 90.0f * 26.0f, size.height / 60.0f * 22.85f)];
    // bottom left
    [path addLineToPoint:CGPointMake(size.width / 90.0f * 26.0f, size.height / 60.0f * 28.4f)];

    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 19.7f, size.height / 60.0f * 29.0f)
            controlPoint1:CGPointMake(size.width / 90.0f * 24.0f, size.height / 60.0f * 28.3f)
            controlPoint2:CGPointMake(size.width / 90.0f * 22.6f, size.height / 60.0f * 28.4f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 19.7f, size.height / 60.0f * 25.4f)
            controlPoint1:CGPointMake(size.width / 90.0f * 19.75f, size.height / 60.0f * 27.2f)
            controlPoint2:CGPointMake(size.width / 90.0f * 19.75f, size.height / 60.0f * 27.2f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 26.0f, size.height / 60.0f * 26.1f)
            controlPoint1:CGPointMake(size.width / 90.0f * 23.6f, size.height / 60.0f * 26.3f)
            controlPoint2:CGPointMake(size.width / 90.0f * 25.0f, size.height / 60.0f * 26.1f)];

    [path addLineToPoint:CGPointMake(size.width / 90.0f * 26.0f, size.height / 60.0f * 22.9f)];
    // middle let
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 21.0f, size.height / 60.0f * 23.6f)
            controlPoint1:CGPointMake(size.width / 90.0f * 24.0f, size.height / 60.0f * 22.9f)
            controlPoint2:CGPointMake(size.width / 90.0f * 22.6f, size.height / 60.0f * 23.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 21.0f, size.height / 60.0f * 20.0f)
            controlPoint1:CGPointMake(size.width / 90.0f * 21.05f, size.height / 60.0f * 21.8f)
            controlPoint2:CGPointMake(size.width / 90.0f * 21.05f, size.height / 60.0f * 21.8f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 26.0f, size.height / 60.0f * 20.7f)
            controlPoint1:CGPointMake(size.width / 90.0f * 23.6f, size.height / 60.0f * 20.8f)
            controlPoint2:CGPointMake(size.width / 90.0f * 25.0f, size.height / 60.0f * 20.6f)];
    // top
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 25.3f, size.height / 60.0f * 16.05f)
            controlPoint1:CGPointMake(size.width / 90.0f * 25.9f, size.height / 60.0f * 19.0f)
            controlPoint2:CGPointMake(size.width / 90.0f * 25.9f, size.height / 60.0f * 18.6f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 28.8f, size.height / 60.0f * 16.05f)
            controlPoint1:CGPointMake(size.width / 90.0f * 27.05f, size.height / 60.0f * 16.15f)
            controlPoint2:CGPointMake(size.width / 90.0f * 27.05f, size.height / 60.0f * 16.15f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 28.0f, size.height / 60.0f * 20.7f)
            controlPoint1:CGPointMake(size.width / 90.0f * 28.2f, size.height / 60.0f * 18.5f)
            controlPoint2:CGPointMake(size.width / 90.0f * 28.2f, size.height / 60.0f * 19.0f)];
    // middle right
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 33.0f, size.height / 60.0f * 20.0f)
            controlPoint1:CGPointMake(size.width / 90.0f * 30.8f, size.height / 60.0f * 20.5f)
            controlPoint2:CGPointMake(size.width / 90.0f * 30.8f, size.height / 60.0f * 20.5f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 33.0f, size.height / 60.0f * 23.6f)
            controlPoint1:CGPointMake(size.width / 90.0f * 32.95f, size.height / 60.0f * 21.8f)
            controlPoint2:CGPointMake(size.width / 90.0f * 32.95f, size.height / 60.0f * 21.8f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 28.0f, size.height / 60.0f * 22.9f)
            controlPoint1:CGPointMake(size.width / 90.0f * 30.5f, size.height / 60.0f * 23.0f)
            controlPoint2:CGPointMake(size.width / 90.0f * 30.1f, size.height / 60.0f * 22.9f)];

    [path addLineToPoint:CGPointMake(size.width / 90.0f * 28.0f, size.height / 60.0f * 26.1f)];

    // bottom right
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 34.3f, size.height / 60.0f * 25.4f)
            controlPoint1:CGPointMake(size.width / 90.0f * 32.0f, size.height / 60.0f * 26.0f)
            controlPoint2:CGPointMake(size.width / 90.0f * 32.0f, size.height / 60.0f * 26.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 34.3f, size.height / 60.0f * 29.0f)
            controlPoint1:CGPointMake(size.width / 90.0f * 34.25f, size.height / 60.0f * 27.2f)
            controlPoint2:CGPointMake(size.width / 90.0f * 34.25f, size.height / 60.0f * 27.2f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 28.0f, size.height / 60.0f * 28.4f)
            controlPoint1:CGPointMake(size.width / 90.0f * 30.5f, size.height / 60.0f * 28.3f)
            controlPoint2:CGPointMake(size.width / 90.0f * 30.5f, size.height / 60.0f * 28.3f)];

    [path addLineToPoint:CGPointMake(size.width / 90.0f * 28.0f, size.height / 60.0f * 40.0f)];
    [path closePath];
    [path fill];

    // draw cloud
    [colorBlue setFill];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 90.0f * 27.0f, size.height / 60.0f * 45.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 17.3f, size.height / 60.0f * 37.0f)
            controlPoint1:CGPointMake(size.width / 90.0f * 23.8f, size.height / 60.0f * 43.5f)
            controlPoint2:CGPointMake(size.width / 90.0f * 20.0f, size.height / 60.0f * 41.3f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 23.2f, size.height / 60.0f * 36.3f)
            controlPoint1:CGPointMake(size.width / 90.0f * 19.1f, size.height / 60.0f * 34.0f)
            controlPoint2:CGPointMake(size.width / 90.0f * 21.4f, size.height / 60.0f * 34.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 30.85f, size.height / 60.0f * 36.3f)
            controlPoint1:CGPointMake(size.width / 90.0f * 25.5f, size.height / 60.0f * 31.7f)
            controlPoint2:CGPointMake(size.width / 90.0f * 28.5f, size.height / 60.0f * 31.7f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 36.61f, size.height / 60.0f * 37.0f)
            controlPoint1:CGPointMake(size.width / 90.0f * 32.6f, size.height / 60.0f * 34.0f)
            controlPoint2:CGPointMake(size.width / 90.0f * 34.6f, size.height / 60.0f * 34.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 90.0f * 27.0f, size.height / 60.0f * 45.0f)
            controlPoint1:CGPointMake(size.width / 90.0f * 33.8f, size.height / 60.0f * 41.3f)
            controlPoint2:CGPointMake(size.width / 90.0f * 30.6f, size.height / 60.0f * 43.2f)];
    [path closePath];
    [path fill];

    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Slovakia.swift
//
//  UIImage+Slovakia.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Slovakia "wikipedia"
     
     Generates the flag of the Slovak Republic.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of the Slovak Republic.
     */
    internal static func flagSlovakia(with size: CGSize) -> UIImage? {
        let colorBlue = UIColor(red: 37.0 / 255.0,
                                green: 74.0 / 255.0,
                                blue: 165.0 / 255.0,
                                alpha: 1.0)
        let colorRed = UIColor(red: 237.0 / 255.0,
                               green: 28.0 / 255.0,
                               blue: 36.0 / 255.0,
                               alpha: 1.0)
        let colorWhite = UIColor.white
        let lineSize = CGSize(width: size.width, height: size.height / 3.0)

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorWhite.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorBlue.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 3.0), size: lineSize))

        colorRed.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 3.0 * 2.0), size: lineSize))

        colorWhite.setStroke()
        colorRed.setFill()
        let path = UIBezierPath()
        // draw shield
        path.move(to: CGPoint(x: size.width / 90.0 * 15.0, y: size.height / 60.0 * 14.5))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 14.5, y: size.height / 60.0 * 28.5),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 14.7, y: size.height / 60.0 * 18.0),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 14.7, y: size.height / 60.0 * 20.0))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 27.0, y: size.height / 60.0 * 45.5),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 14.8, y: size.height / 60.0 * 36.3),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 18.1, y: size.height / 60.0 * 40.8))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 39.5, y: size.height / 60.0 * 28.5),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 40.1, y: size.height / 60.0 * 38.8),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 39.0, y: size.height / 60.0 * 31.0))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 38.96, y: size.height / 60.0 * 14.5),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 39.4, y: size.height / 60.0 * 20.0),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 39.3, y: size.height / 60.0 * 18.0))
        path.addLine(to: CGPoint(x: size.width / 90.0 * 15.0, y: size.height / 60.0 * 14.5))
        path.close()
        path.lineWidth = size.width / 90.0 * (9.0 / 10.0)
        path.fill()
        path.stroke()

        // draw central line
        colorWhite.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 90.0 * 26.0, y: size.height / 60.0 * 40.0))
        path.addLine(to: CGPoint(x: size.width / 90.0 * 26.0, y: size.height / 60.0 * 22.85))
        // bottom left
        path.addLine(to: CGPoint(x: size.width / 90.0 * 26.0, y: size.height / 60.0 * 28.4))

        path.addCurve(to: CGPoint(x: size.width / 90.0 * 19.7, y: size.height / 60.0 * 29.0),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 24.0, y: size.height / 60.0 * 28.3),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 22.6, y: size.height / 60.0 * 28.4))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 19.7, y: size.height / 60.0 * 25.4),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 19.75, y: size.height / 60.0 * 27.2),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 19.75, y: size.height / 60.0 * 27.2))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 26.0, y: size.height / 60.0 * 26.1),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 23.6, y: size.height / 60.0 * 26.3),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 25.0, y: size.height / 60.0 * 26.1))

        path.addLine(to: CGPoint(x: size.width / 90.0 * 26.0, y: size.height / 60.0 * 22.9))
        // middle let
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 21.0, y: size.height / 60.0 * 23.6),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 24.0, y: size.height / 60.0 * 22.9),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 22.6, y: size.height / 60.0 * 23.0))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 21.0, y: size.height / 60.0 * 20.0),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 21.05, y: size.height / 60.0 * 21.8),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 21.05, y: size.height / 60.0 * 21.8))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 26.0, y: size.height / 60.0 * 20.7),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 23.6, y: size.height / 60.0 * 20.8),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 25.0, y: size.height / 60.0 * 20.6))
        // top
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 25.3, y: size.height / 60.0 * 16.05),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 25.9, y: size.height / 60.0 * 19.0),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 25.9, y: size.height / 60.0 * 18.6))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 28.8, y: size.height / 60.0 * 16.05),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 27.05, y: size.height / 60.0 * 16.15),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 27.05, y: size.height / 60.0 * 16.15))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 28.0, y: size.height / 60.0 * 20.7),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 28.2, y: size.height / 60.0 * 18.5),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 28.2, y: size.height / 60.0 * 19.0))
        // middle right
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 33.0, y: size.height / 60.0 * 20.0),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 30.8, y: size.height / 60.0 * 20.5),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 30.8, y: size.height / 60.0 * 20.5))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 33.0, y: size.height / 60.0 * 23.6),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 32.95, y: size.height / 60.0 * 21.8),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 32.95, y: size.height / 60.0 * 21.8))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 28.0, y: size.height / 60.0 * 22.9),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 30.5, y: size.height / 60.0 * 23.0),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 30.1, y: size.height / 60.0 * 22.9))

        path.addLine(to: CGPoint(x: size.width / 90.0 * 28.0, y: size.height / 60.0 * 26.1))

        // bottom right
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 34.3, y: size.height / 60.0 * 25.4),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 32.0, y: size.height / 60.0 * 26.0),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 32.0, y: size.height / 60.0 * 26.0))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 34.3, y: size.height / 60.0 * 29.0),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 34.25, y: size.height / 60.0 * 27.2),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 34.25, y: size.height / 60.0 * 27.2))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 28.0, y: size.height / 60.0 * 28.4),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 30.5, y: size.height / 60.0 * 28.3),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 30.5, y: size.height / 60.0 * 28.3))

        path.addLine(to: CGPoint(x: size.width / 90.0 * 28.0, y: size.height / 60.0 * 40.0))
        path.close()
        path.fill()

        // draw cloud
        colorBlue.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 90.0 * 27.0, y: size.height / 60.0 * 45.0))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 17.3, y: size.height / 60.0 * 37.0),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 23.8, y: size.height / 60.0 * 43.5),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 20.0, y: size.height / 60.0 * 41.3))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 23.2, y: size.height / 60.0 * 36.3),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 19.1, y: size.height / 60.0 * 34.0),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 21.4, y: size.height / 60.0 * 34.0))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 30.85, y: size.height / 60.0 * 36.3),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 25.5, y: size.height / 60.0 * 31.7),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 28.5, y: size.height / 60.0 * 31.7))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 36.61, y: size.height / 60.0 * 37.0),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 32.6, y: size.height / 60.0 * 34.0),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 34.6, y: size.height / 60.0 * 34.0))
        path.addCurve(to: CGPoint(x: size.width / 90.0 * 27.0, y: size.height / 60.0 * 45.0),
                      controlPoint1: CGPoint(x: size.width / 90.0 * 33.8, y: size.height / 60.0 * 41.3),
                      controlPoint2: CGPoint(x: size.width / 90.0 * 30.6, y: size.height / 60.0 * 43.2))
        path.close()
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
